// appium-remote-debugger/build/lib/protocol/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
exports.getProtocolCommand = getProtocolCommand;

require("source-map-support/register");

const OBJECT_GROUP = 'console';
const COMMANDS = {
  'ApplicationCache.enable': [],
  'ApplicationCache.getFramesWithManifests': [],
  'Canvas.enable': [],
  'Console.disable': [],
  'Console.enable': [],
  'Console.getLoggingChannels': [],
  'Console.setLoggingChannelLevel': ['source', 'level'],
  'CSS.enable': [],
  'Database.enable': [],
  'Debugger.enable': [],
  'Debugger.setAsyncStackTraceDepth': ['depth'],
  'Debugger.setBreakpointsActive': ['active'],
  'Debugger.setPauseForInternalScripts': ['shouldPause'],
  'Debugger.setPauseOnAssertions': ['enabled'],
  'Debugger.setPauseOnExceptions': ['state'],
  'DOM.getDocument': [],
  'DOMStorage.enable': [],
  'Heap.enable': [],
  'Heap.gc': [],
  'IndexedDB.enable': [],
  'Inspector.enable': [],
  'Inspector.initialized': [],
  'LayerTree.enable': [],
  'Memory.enable': [],
  'Network.disable': [],
  'Network.enable': [],
  'Network.setResourceCachingDisabled': ['disabled'],
  'Page.deleteCookie': ['cookieName', 'url'],
  'Page.enable': [],
  'Page.getCookies': ['urls'],
  'Page.getResourceTree': [],
  'Page.navigate': ['url'],
  'Runtime.awaitPromise': ['promiseObjectId', 'returnByValue', 'generatePreview', 'saveResult'],
  'Runtime.callFunctionOn': ['objectId', 'functionDeclaration', 'arguments', 'returnByValue'],
  'Runtime.enable': [],
  'Runtime.evaluate': ['expression', 'returnByValue', 'contextId'],
  'Target.exists': [],
  'Timeline.setAutoCaptureEnabled': ['enabled'],
  'Timeline.setInstruments': ['instruments'],
  'Timeline.start': [],
  'Timeline.stop': [],
  'Worker.enable': []
};

function getCommand(id, method, params = {}) {
  return {
    id,
    method,
    params: Object.assign({
      objectGroup: OBJECT_GROUP,
      includeCommandLineAPI: true,
      doNotPauseOnExceptionsAndMuteConsole: false,
      emulateUserGesture: false,
      generatePreview: false,
      saveResult: false
    }, params)
  };
}

function getProtocolCommand(id, method, opts = {}) {
  const paramNames = COMMANDS[method];

  if (!paramNames) {
    throw new Error(`Unknown command: '${method}'`);
  }

  const params = paramNames.reduce(function (params, name) {
    params[name] = opts[name];
    return params;
  }, {});
  return getCommand(id, method, params);
}

var _default = getProtocolCommand;
exports.default = _default;require('source-map-support').install();
