// appium-ios-device/build/lib/instrument/transformer/dtx-decode.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DTXDecoder = void 0;

require("source-map-support/register");

var _stream = _interopRequireDefault(require("stream"));

var _headers = require("../headers");

class DTXDecoder extends _stream.default.Transform {
  constructor() {
    super({
      objectMode: true
    });
    this._dtxManager = {};
    this.buffer = Buffer.allocUnsafe(0);
    this.cursor = 0;
    this.header = undefined;
  }

  _transform(data, encoding, onData) {
    this._decode(data);

    onData();
  }

  _decode(data) {
    this.buffer = Buffer.concat([this.buffer.slice(this.cursor, this.buffer.length), data]);
    this.cursor = 0;

    while (this.cursor < this.buffer.length) {
      const magic = this._recv(_headers.DTX_MESSAGE_HEADER_MAGIC_LEN);

      if (!magic) {
        return;
      }

      this.cursor -= _headers.DTX_MESSAGE_HEADER_MAGIC_LEN;

      if (magic && magic.readUInt32LE(0) === _headers.DTX_MESSAGE_HEADER_MAGIC) {
        const headerBuffer = this._recv(_headers.DTX_MESSAGE_HEADER_LENGTH);

        if (!headerBuffer) {
          return;
        }

        this.header = _headers.DTXMessageHeader.parse(headerBuffer);

        if (this.header.fragmentId === 0) {
          if (!(this.header.channel in this._dtxManager)) {
            this._dtxManager[this.header.channel] = {
              headerBuffer,
              payloadBuffer: new Buffer.allocUnsafe(0)
            };
          }

          if (this.header.fragmentCount > 1) {
            continue;
          }
        }
      }

      const bodyBuffer = this._recv(this.header.payloadLength);

      if (!bodyBuffer) {
        return;
      }

      if (this._dtxManager[this.header.channel]) {
        this._dtxManager[this.header.channel].payloadBuffer = Buffer.concat([this._dtxManager[this.header.channel].payloadBuffer, bodyBuffer]);
      }

      if (this.header.fragmentId === this.header.fragmentCount - 1) {
        data = this._dtxManager[this.header.channel];
        delete this._dtxManager[this.header.channel];

        if (data) {
          this.push(_headers.DTXMessage.parse(data.headerBuffer, data.payloadBuffer));
        }
      }
    }

    this.cursor = 0;
    this.buffer = Buffer.allocUnsafe(0);
  }

  _recv(length) {
    if (this.buffer.length < this.cursor + length) {
      return null;
    }

    const buf = this.buffer.slice(this.cursor, this.cursor + length);
    this.cursor += length;
    return buf;
  }

}

exports.DTXDecoder = DTXDecoder;require('source-map-support').install();
