// appium-ios-device/build/lib/mcinstall/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.MC_INSTALL_SERVICE_NAME = exports.MCInstallProxyService = void 0;

require("source-map-support/register");

var _baseService = require("../base-service");

var _support = require("@appium/support");

const MC_INSTALL_SERVICE_NAME = 'com.apple.mobile.MCInstall';
exports.MC_INSTALL_SERVICE_NAME = MC_INSTALL_SERVICE_NAME;
const ACKNOWLEDGED = 'Acknowledged';

function checkACK(res, name) {
  if (res.Status !== ACKNOWLEDGED) {
    throw new Error(`${name} error: ${JSON.stringify(res)}`);
  }

  return res;
}

class MCInstallProxyService extends _baseService.BaseServicePlist {
  constructor(socketClient) {
    super(socketClient);
  }

  async getProfileList() {
    const res = await this._plistService.sendPlistAndReceive({
      RequestType: 'GetProfileList'
    });
    return checkACK(res, 'getProfileList');
  }

  async installProfile(path) {
    const payload = await _support.fs.readFile(path);
    const res = await this._plistService.sendPlistAndReceive({
      'RequestType': 'InstallProfile',
      'Payload': payload
    });
    return checkACK(res, 'installProfile');
  }

  async removeProfile(ident) {
    const profiles = await this.getProfileList();
    const meta = profiles.ProfileMetadata[ident];

    if (!meta) {
      throw new Error(`not find installed profile ident:${ident}`);
    }

    const data = _support.plist.createBinaryPlist({
      'PayloadType': 'Configuration',
      'PayloadIdentifier': ident,
      'PayloadUUID': meta.PayloadUUID,
      'PayloadVersion': meta.PayloadVersion
    });

    const res = await this._plistService.sendPlistAndReceive({
      'RequestType': 'RemoveProfile',
      'ProfileIdentifier': data
    });
    return checkACK(res, 'removeProfile');
  }

}

exports.MCInstallProxyService = MCInstallProxyService;
var _default = MCInstallProxyService;
exports.default = _default;require('source-map-support').install();
