// appium-xcuitest-driver/build/lib/commands/biometric.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.helpers = exports.default = exports.commands = void 0;

require("source-map-support/register");

let commands = {},
    helpers = {},
    extensions = {};
exports.helpers = helpers;
exports.commands = commands;

function assertIsSimulator(driver) {
  if (!driver.isSimulator()) {
    throw new Error('Biometric enrollment can only be performed on Simulator');
  }
}

commands.mobileEnrollBiometric = async function mobileEnrollBiometric(opts = {}) {
  const {
    isEnabled = true
  } = opts;
  assertIsSimulator(this);
  await this.opts.device.enrollBiometric(isEnabled);
};

commands.mobileSendBiometricMatch = async function mobileSendBiometricMatch(opts = {}) {
  const {
    match = true,
    type = 'touchId'
  } = opts;
  assertIsSimulator(this);
  await this.opts.device.sendBiometricMatch(match, type);
};

commands.mobileIsBiometricEnrolled = async function mobileIsBiometricEnrolled() {
  assertIsSimulator(this);
  return await this.opts.device.isBiometricEnrolled();
};

Object.assign(extensions, commands, helpers);
var _default = extensions;
exports.default = _default;require('source-map-support').install();
