// appium-idb/build/lib/idb.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.IDB = void 0;

require("source-map-support/register");

var _lodash = _interopRequireDefault(require("lodash"));

var _index = _interopRequireDefault(require("./tools/index.js"));

var _helpers = require("./helpers");

const DEFAULT_OPTS = {
  udid: null,
  executable: {
    path: _helpers.IDB_EXECUTABLE,
    port: null,
    grpcPort: null,
    defaultArgs: []
  },
  logLevel: null,
  companion: {
    path: _helpers.IDB_COMPANION_EXECUTABLE,
    port: null,
    grpcPort: null,
    logPath: null
  },
  execTimeout: _helpers.DEFAULT_IDB_EXEC_TIMEOUT,
  verbose: false
};

class IDB {
  constructor(opts = {}) {
    Object.assign(this, opts);

    _lodash.default.defaultsDeep(this, _lodash.default.cloneDeep(DEFAULT_OPTS));

    if (!this.udid) {
      throw new Error(`UDID must be set for idb`);
    }

    this.executable.defaultArgs.push('--udid', this.udid);

    if (this.logLevel) {
      this.executable.defaultArgs.push('--log', this.logLevel);
    }
  }

}

exports.IDB = IDB;

for (const [fnName, fn] of _lodash.default.toPairs(_index.default)) {
  IDB.prototype[fnName] = fn;
}

var _default = IDB;
exports.default = _default;require('source-map-support').install();
