// @appium/test-support/build/lib/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "fakeTime", {
  enumerable: true,
  get: function () {
    return _timeUtils.fakeTime;
  }
});
Object.defineProperty(exports, "stubEnv", {
  enumerable: true,
  get: function () {
    return _envUtils.stubEnv;
  }
});
Object.defineProperty(exports, "stubLog", {
  enumerable: true,
  get: function () {
    return _logUtils.stubLog;
  }
});
Object.defineProperty(exports, "verifyMocks", {
  enumerable: true,
  get: function () {
    return _mockUtils.verifyMocks;
  }
});
Object.defineProperty(exports, "verifySandbox", {
  enumerable: true,
  get: function () {
    return _sandboxUtils.verifySandbox;
  }
});
Object.defineProperty(exports, "withMocks", {
  enumerable: true,
  get: function () {
    return _mockUtils.withMocks;
  }
});
Object.defineProperty(exports, "withSandbox", {
  enumerable: true,
  get: function () {
    return _sandboxUtils.withSandbox;
  }
});

require("source-map-support/register");

var _envUtils = require("./env-utils");

var _logUtils = require("./log-utils");

var _timeUtils = require("./time-utils");

var _mockUtils = require("./mock-utils");

var _sandboxUtils = require("./sandbox-utils");

require("./unhandled-rejection");
