// @appium/test-support/build/lib/mock-utils.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.verifyMocks = verifyMocks;
exports.withMocks = withMocks;

require("source-map-support/register");

var _sinon = _interopRequireDefault(require("sinon"));

var _lodash = _interopRequireDefault(require("lodash"));

let SANDBOX = Symbol();

function withMocks(libs, fn) {
  return () => {
    const mocks = {
      verify() {
        this.sandbox.verify();
      },

      get sandbox() {
        return this[SANDBOX];
      },

      set sandbox(sandbox) {
        this[SANDBOX] = sandbox;
      }

    };
    beforeEach(function beforeEach() {
      mocks[SANDBOX] = _sinon.default.createSandbox();

      for (let [key, value] of _lodash.default.toPairs(libs)) {
        mocks[key] = mocks.sandbox.mock(value);
      }
    });
    afterEach(function afterEach() {
      mocks.sandbox.restore();
    });
    fn(mocks);
  };
}

function verifyMocks(mocks) {
  mocks.sandbox.verify();
}
