// appium-ios-simulator/build/lib/calendar.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

require("source-map-support/register");

var _lodash = _interopRequireDefault(require("lodash"));

var _logger = _interopRequireDefault(require("./logger"));

var _tccDb = _interopRequireDefault(require("./tcc-db"));

class Calendar {
  constructor(xcodeVersion, sharedResourcesDir) {
    this.xcodeVersion = xcodeVersion;
    this.tccDb = new _tccDb.default(xcodeVersion, sharedResourcesDir);
  }

  async getCalendarRowCount(bundleID) {
    const count = await this.tccDb.execQuery(`SELECT count(*) FROM access WHERE client='?' AND service='kTCCServiceCalendar';`, bundleID);
    return parseInt(count.split('=')[1], 10);
  }

  async hasCalendarAccess(bundleID) {
    const count = await this.tccDb.execQuery(`SELECT count(*) FROM access WHERE client='?' AND allowed=1 AND service='kTCCServiceCalendar';`, bundleID);
    return parseInt(count.split('=')[1], 10) > 0;
  }

  async enableCalendarAccess(bundleID) {
    let query;

    if ((await this.getCalendarRowCount(bundleID)) > 0) {
      query = `UPDATE 'access' SET
        service='kTCCServiceCalendar',
        client_type=0,
        allowed=1,
        prompt_count=1,
        csreq=0
      WHERE client='?' AND service='kTCCServiceCalendar'`;
    } else {
      let count = this.columnCount;

      if (!_lodash.default.isNumber(this.columnCount)) {
        try {
          const stdout = await this.tccDb.execQuery(`pragma table_info('access')`);
          count = this.columnCount = stdout.trim().split('cid').filter(record => !_lodash.default.isEmpty(record)).length;
        } catch (err) {
          count = this.xcodeVersion.major < 10 ? 7 : 11;

          _logger.default.warn(`Unable to find Calendar access column count: ${err.message}`);
        }
      }

      query = `INSERT INTO 'access' VALUES ('kTCCServiceCalendar', '?', 0, 1, 1, ${_lodash.default.range(0, count - 5, 0).join(', ')});`;
    }

    await this.tccDb.execQuery(query, bundleID);
  }

  async disableCalendarAccess(bundleID) {
    if ((await this.getCalendarRowCount(bundleID)) > 0) {
      await this.tccDb.execQuery(`DELETE FROM 'access' WHERE client='?' AND service='kTCCServiceCalendar'`, bundleID);
    }
  }

}

var _default = Calendar;
exports.default = _default;require('source-map-support').install();
